[core] Fix performance regression in single_client_tasks_and_get_batch

vitsai opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

The `single_client_tasks_and_get_batch` benchmark saw a ~0.5-1k tasks/s average regression (2k tasks/s on a local machine) due to https://github.com/ray-project/ray/pull/38323, which changed some tracing logic to unconditionally change the signature of every remote function to accomodate tracing during _inject_tracing_into_function.

Make the signature change conditional again, but move it to the execution portion of `RemoteFunction` rather than the definition. Also make sure the injection only happens once even when the remote function is executed multiple times.

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

https://github.com/ray-project/ray/issues/39259

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

rkooo567 commented:
Running microbenchmark here https://buildkite.com/ray-project/release-tests-pr/builds/52417.

@vitsai can you let me know the number after this PR from this test after it is completed?

rkooo567 commented on python/ray/util/tracing/tracing_helper.py:
```diff
@@ -328,7 +328,9 @@ def _inject_tracing_into_function(function):
     Use the provided trace context from kwargs.
     """
-    # Add _ray_trace_ctx to function signature
+    if not _is_tracing_enabled():
```
This change means that the change up there in remote_function.py is just for the optimization when tracing is enabled right? Just with this fix, if the tracing is not enabled, there would be no regression

vitsai replied:
Just with this, there would be no regression, but there would be a correctness issue because `_is_tracing_enabled()` is not accurate at the time this function was originally being called in `remote_function.py` (needs to be after `init()`)

rkooo567 commented on python/ray/remote_function.py:
```diff
@@ -116,14 +116,12 @@ def __init__(
         self._language = language
-        self._function = _inject_tracing_into_function(function)
+        self._function = function
+        self._injected_tracing = False
```
nit, but let's just use

```python3
self._function_signature = None
```

And use 

```python3
        if self._function_signature is None:
            self._function = _inject_tracing_into_function(self._function)
            self._function_signature = ray._private.signature.extract_signature(
                self._function
            )
```
below?

rkooo567 commented on python/ray/remote_function.py:
```diff
@@ -254,6 +252,15 @@ def _remote(self, args=None, kwargs=None, **task_options):
         worker = ray._private.worker.global_worker
         worker.check_connected()
 
+        # We cannot do this when the function is first defined, because we need
```
Does this mean the before this change, it was already broken? I think when the remote function is first created, ray should automatically call ray.init

edoakes replied:
ray calls ray.init when the function is first invoked, not defined

rkooo567 reviewed (approved):
Looks good to me. Nit comments

edoakes commented on python/ray/remote_function.py:
```diff
@@ -254,6 +252,15 @@ def _remote(self, args=None, kwargs=None, **task_options):
 
+        # We cannot do this when the function is first defined, because we need
+        # ray.init() to have been called when this executes
+        if not self._injected_tracing:
+            self._function = _inject_tracing_into_function(self._function)
+            self._function_signature = ray._private.signature.extract_signature(
+                self._function
+            )
+            self._injected_tracing = True
```
the ray API needs to be thread-safe, is there any concern of a race condition here if multiple threads invoke a function concurrently?

vitsai replied:
Actually no `_inject` is not idempotent so

vitsai commented:
`single_client_tasks_and_get_batch = [12.225312658308761, 0.38310197760588016]` for [microbenchmark release test](https://buildkite.com/ray-project/release-tests-pr/builds/52552#018a7168-8a87-45c8-8193-aa9f833d433b) after addressing comments (does not include latest commit, which only changes 1 line of test logic)

edoakes commented on python/ray/remote_function.py:
```diff
@@ -254,6 +264,18 @@ def _remote(self, args=None, kwargs=None, **task_options):
+        if self._function_signature is None:
+            # Since variable assignment is atomic, we don't need to acquire the
+            # lock before checking _function_signature.
+            self._inject_mut.acquire()
+            self._function = _inject_tracing_into_function(self._function)
+            self._function_signature = ray._private.signature.extract_signature(
+                self._function
+            )
+            self._inject_mut.release()
```
should use `with self._inject_mut:` here instead of acquire/release so the lock is released if an exception is raised in either of these calls (also just standard pattern)

edoakes commented on python/ray/remote_function.py:
```diff
@@ -116,14 +117,13 @@ def __init__(
-        self._function = _inject_tracing_into_function(function)
+        self._function = function
+        self._function_signature = None
+        self._inject_mut = Lock()
```
nit: prefer `_inject_lock` to match type

also, add a small comment describing what this guards?

edoakes reviewed (commented):
One small comment otherwise lgtm

rkooo567 commented on python/ray/remote_function.py:
```diff
@@ -254,6 +265,17 @@ def _remote(self, args=None, kwargs=None, **task_options):
+        if self._function_signature is None:
+            # Since variable assignment is atomic, we don't need to acquire the
+            # lock before checking _function_signature.
+            with self._inject_lock:
```
actually can you do 


```suggestion
            with self._inject_lock:
                if self._function_signature is None:
                    ...
```

Otherwise, there's race condition between checking function signature and lock

vitsai replied:
there shouldn't be a race because variable assignment is atomic, though?

vitsai replied:
oops no data race but still should have checked again inside the with

rkooo567 commented:
Lmk when tests pass! Seems good to merge

vitsai commented:
Tests look good, seems like microbenchmark ran into some infra failure so running it again now. On my machine, I did see the throughput go back up to ~13k/s from ~11k/s on the latest commit, so personally don't think we have to wait.

rkooo567 commented:
You machine has a different environment from microbenchmark, so we shouldn't rely on that result. Let me rerun tests

rkooo567 commented:
Try again here; https://buildkite.com/ray-project/release-tests-pr/builds/52670

rkooo567 commented:
wait, actually I realized it succeeded? https://buildkite.com/ray-project/release-tests-pr/builds/52417#018a6f01-015c-4d80-8f8c-f37dc1a0a686

rkooo567 commented:
=



single_client_tasks_and_get_batch = [11.222801689347772, 0.4976784102185062]
 

<br class="Apple-interchange-newline



single_client_tasks_and_get_batch = [11.222801689347772, 0.4976784102185062]
 

<br class="Apple-interchange-newline">

vitsai commented:
(probably because I retried at the time of failure in the last comment)
